### PR TITLE
oauth2: added refresh token generation for password grant type

### PR DESCRIPTION
**Pull request by jrossiter**
I'm currently porting over some old authentication services with fosite and have ran into the issue specified here: https://github.com/ory-am/fosite/issues/78

Where internal clients will make use of the password grant type along with the access / refresh tokens.

There was a comment suggesting that this particular feature be added as a separate PR than the issue raised; and since it's been around a month with no updates I decided to give it a go.

Please let me know if anything doesn't meet the RFC or needs changes.

### CODE REVIEW

**Comment by arekkas** [comment]
Awesome, thanks for this! I will take a closer look tonight!

File: fosite-example/main.go [hunk 1]
```diff
@@ -131,6 +131,11 @@ func tokenEndpoint(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	// Grant requested scopes
```
Comment (arekkas) [inline fosite-example/main.go]: Why did you add this? Is it to get the resource owner credentials grant working with scopes?

File: handler/oauth2/flow_resource_owner_test.go [hunk 2]
```diff
@@ -111,6 +113,17 @@ func TestResourceOwnerFlow_PopulateTokenEndpointResponse(t *testing.T) {
...
+		{
+			description: "should pass - offline scope",
+			setup: func() {
+				areq.GrantTypes = fosite.Arguments{"password"}
+				areq.GrantScope("offline")
+				rtstr.EXPECT().GenerateRefreshToken(nil, areq).Return("tokenfoo.bar", "bar", nil)
+				store.EXPECT().CreateRefreshTokenSession(nil, "bar", areq).Return(nil)
+				chgen.EXPECT().GenerateAccessToken(nil, areq).Return("tokenfoo.bar", "bar", nil)
+				store.EXPECT().CreateAccessTokenSession(nil, "bar", areq).Return(nil)
+			},
```
Comment (arekkas) [inline handler/oauth2/flow_resource_owner_test.go]: Should we add a test here if `"refresh_token"` is part of the response? What do you think?

File: fosite-example/main.go (hunk 1 above)
Comment (jrossiter) [inline fosite-example/main.go]: In `ResourceOwnerPasswordCredentialsGrantHandler.PopulateTokenEndpointResponse()` 
I'm using `if requester.GetGrantedScopes().Has("offline")` to determine if they need a refresh token but it seems like this wasn't set anywhere. 
I compared it to the authorize endpoint which set the scopes like this.

The NewAccessRequest seems to already throw an error if a requested scope isn't allowed by a client so I thought this may be safe to do here.

What do you think?

File: handler/oauth2/flow_resource_owner_test.go (hunk 2 above)
Comment (jrossiter) [inline handler/oauth2/flow_resource_owner_test.go]: Good idea; I'll add to the tests to check for `refresh_token` with and without the `offline` scope.

Comment (jrossiter) [inline handler/oauth2/flow_resource_owner_test.go]: I've merged the new checks.

File: fosite-example/main.go (hunk 1 above)
Comment (arekkas) [inline fosite-example/main.go]: Yup sounds good!

File: handler/oauth2/flow_resource_owner_test.go [hunk 3]
```diff
@@ -107,14 +111,34 @@ func TestResourceOwnerFlow_PopulateTokenEndpointResponse(t *testing.T) {
...
+				chgen.EXPECT().GenerateAccessToken(nil, areq).Return(mockAT, "bar", nil)
 				store.EXPECT().CreateAccessTokenSession(nil, "bar", areq).Return(nil)
 			},
 		},
 	} {
 		c.setup()
 		err := h.PopulateTokenEndpointResponse(nil, httpreq, areq, aresp)
 		assert.True(t, errors.Cause(err) == c.expectErr, "(%d) %s\n%s\n%s", k, c.description, err, c.expectErr)
+
+		if err == nil {
```
Comment (arekkas) [inline handler/oauth2/flow_resource_owner_test.go]: You could probably solve this with less coupling of test cases and test executing by introducing a `expect func()` field to the case struct and check here if it's nil and if not execute it. The passing test case could then define a check function that executes the assertion logic. You can check this code for reference: https://github.com/ory-am/fosite/blob/bef61973fdee1a18aedba4e42a1d8977c3f8cc1c/handler/oauth2/flow_refresh_test.go#L92-L98

What do you think of this approach?

Comment (jrossiter) [inline handler/oauth2/flow_resource_owner_test.go]: I like this approach. Just updated the branch.

**Comment by 256dpi** [comment]
Very cool indeed!

AFAIK, the documentation does not explicitly mention that the "offline" scope is used to determine that a client can use refresh tokens. Could you guys maybe add some documentation about how the mechanism is working in fosite?

**Comment by arekkas** [comment]
[this](https://github.com/ory-am/fosite#scopes) could be the right place for it, do you want to create a PR @256dpi ? :)

**Comment by 256dpi** [comment]
@arekkas I would love to, unfortunately I'm not aware of all the details of the implementation. Therefore someone that already read into that particular code section might be better suited for the job. ;)

**Comment by arekkas** [comment]
Added: https://github.com/ory-am/fosite/commit/60a767221625d0f6541f203e41a7ef20a1782eb0

I promise to take the time at some point to write some docs on gitbook.

**Comment by jrossiter** [comment]
Thanks!
